winml/lib/Api.Ort/OnnxruntimeEngineBuilder.cpp

```diff
@@ -0,0 +1,74 @@
+#include "pch.h"
+
+#include "OnnxruntimeEngine.h"
+#include "OnnxruntimeEngineBuilder.h"
+#include "OnnxruntimeCpuSessionBuilder.h"
+
+#ifdef USE_DML
+#include "OnnxruntimeDmlSessionBuilder.h"
+#endif
+
+using namespace WinML;
+
+HRESULT OnnxruntimeEngineBuilder::RuntimeClassInitialize(OnnxruntimeEngineFactory* engine_factory) {
+  engine_factory_ = engine_factory;
+  return S_OK;
+}
+
+STDMETHODIMP OnnxruntimeEngineBuilder::CreateEngine(Windows::AI::MachineLearning::IEngine** out) {
+  auto ort_api = engine_factory_->UseOrtApi();
+
+  Microsoft::WRL::ComPtr<IOrtSessionBuilder> onnxruntime_session_builder;
+
+  if (device_ == nullptr) {
+    RETURN_IF_FAILED(Microsoft::WRL::MakeAndInitialize<OnnxruntimeCpuSessionBuilder>(&onnxruntime_session_builder, engine_factory_.Get()));
+  } else {
+#ifdef USE_DML
+    RETURN_IF_FAILED(Microsoft::WRL::MakeAndInitialize<OnnxruntimeDmlSessionBuilder>(&onnxruntime_session_builder, engine_factory_.Get(), device_.Get(), queue_.Get()));
+#endif
+  }
+
+  OrtSessionOptions* ort_options;
+  RETURN_IF_FAILED(onnxruntime_session_builder->CreateSessionOptions(&ort_options));
+  auto session_options = UniqueOrtSessionOptions(ort_options, ort_api->ReleaseSessionOptions);
+  
+  if (batch_size_override_.has_value()) {
+    constexpr const char* DATA_BATCH = "DATA_BATCH";
+    ort_api->AddFreeDimensionOverride(session_options.get(), DATA_BATCH, batch_size_override_.value());
+  }
```

Comment:
Can sessionoption configuration be entirely done by engine builders?

Reply:
I would have to pass down properties into each builder for common properties.. which could make the design cumbersome... leaving as is and can revisit in the future.

---
In reply to: [370876589](https://github.com/microsoft/onnxruntime/pull/2891#discussion_r370876589) [](ancestors = 370876589)